### PR TITLE
Add unit tests for com.github.neatlife.jframework.util.PrecisionUtil

### PR DESCRIPTION
**Pull request by Braavos6**
I've analysed your codebase and noticed that `com.github.neatlife.jframework.util.PrecisionUtil` is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

**Review by petersear — APPROVED** [review]

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        Assert.assertNull(PrecisionUtil.minArr(null, "1", "2", "3"));
+        Assert.assertEquals("1", PrecisionUtil.minArr("2", "4", "1", "3"));
+    }
+
+    @Test
+    public void testDoubleRound() {
+        Assert.assertEquals(3.0, PrecisionUtil.round(2.567, 0), 0);
+        Assert.assertEquals(2.6, PrecisionUtil.round(2.567, 1), 0);
+        Assert.assertEquals(2.57, PrecisionUtil.round(2.567, 2), 0);
+        Assert.assertEquals(2.567, PrecisionUtil.round(2.567, 3), 0);
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:143]: Add test with precision `4` (but still using `2.567`)

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Test
+    public void testStringMaxArr() {
+        Assert.assertNull(PrecisionUtil.max(null, "1", "2", "3"));
+        Assert.assertEquals("4", PrecisionUtil.maxArr("2", "4", "1", "3"));
+    }
+
+    @Test
+    public void testStringMinArr() {
+        Assert.assertNull(PrecisionUtil.minArr(null, "1", "2", "3"));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:132]: Add a test where `null` is not the first element, and one with an empty string(s) included

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    @Test
+    public void testBigDecimalMinArr() {
+        Assert.assertEquals(BigDecimal.ZERO, PrecisionUtil.minArr(
+                BigDecimal.ONE, BigDecimal.ZERO, BigDecimal.TEN));
+    }
+
+    @Test
+    public void testStringMaxArr() {
+        Assert.assertNull(PrecisionUtil.max(null, "1", "2", "3"));
+        Assert.assertEquals("4", PrecisionUtil.maxArr("2", "4", "1", "3"));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:125]: Add a test where `null` is not the first element, and one with an empty string(s) included

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import java.math.BigDecimal;
+
+public class PrecisionUtilTest {
+
+    @Test
+    public void testBigDecimalAdd() {
+        Assert.assertEquals(BigDecimal.ONE,
+                PrecisionUtil.add(null, BigDecimal.ONE));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Add test where neither param is null, and in both tests with nulls change the non-null param to something other than `ONE` so it's clear that the non-null param is not the source of the return value.

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        Assert.assertEquals(BigDecimal.ONE,
+                PrecisionUtil.subtract(BigDecimal.ONE, null));
+    }
+
+    @Test
+    public void testBigDecimalMultiply() {
+        Assert.assertEquals(new BigDecimal("10"),
+                PrecisionUtil.multiply(null, BigDecimal.TEN));
+        Assert.assertEquals(BigDecimal.ONE,
+                PrecisionUtil.multiply(BigDecimal.ONE, null));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Add test where neither param is null, and in the second test with null change the non-null param to something other than `ONE` so it's clear that the non-null param is not the source of the return value.

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    @Test
+    public void testBigDecimalDivide() {
+        Assert.assertEquals(new BigDecimal("0.10"),
+                PrecisionUtil.divide(BigDecimal.ONE, BigDecimal.TEN));
+        Assert.assertEquals(new BigDecimal("0.10"),
+                PrecisionUtil.divide(BigDecimal.ONE, BigDecimal.TEN));
+        Assert.assertEquals(new BigDecimal("0"),
+                PrecisionUtil.divide(null, BigDecimal.TEN, 1));
+        Assert.assertEquals(new BigDecimal("1.0"),
+                PrecisionUtil.divide(BigDecimal.ONE, null, 1));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: In this test change the first param to something other than `ONE` so it's clear that the non-null param is not the source of the return value.

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Test
+    public void testBigDecimalDivide() {
+        Assert.assertEquals(new BigDecimal("0.10"),
+                PrecisionUtil.divide(BigDecimal.ONE, BigDecimal.TEN));
+        Assert.assertEquals(new BigDecimal("0.10"),
+                PrecisionUtil.divide(BigDecimal.ONE, BigDecimal.TEN));
+        Assert.assertEquals(new BigDecimal("0"),
+                PrecisionUtil.divide(null, BigDecimal.TEN, 1));
+        Assert.assertEquals(new BigDecimal("1.0"),
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: In all tests use `BigDecimal.ONE` instead of `new BigDecimal("1.0")` (same for `TEN` and `ZERO` wherever possible)

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    @Test
+    public void testStringMultiply() {
+        Assert.assertEquals("1", PrecisionUtil.multiply("1", null));
+        Assert.assertEquals("2", PrecisionUtil.multiply(null, "2"));
+    }
+
+    @Test
+    public void testStringDivide() {
+        Assert.assertEquals("1.00", PrecisionUtil.divide("1", null));
+        Assert.assertEquals("0", PrecisionUtil.divide(null, "2"));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:77]: Add a test in each of these four `testStringX` tests that does not have either parameter `null`.

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    @Test
+    public void testBigDecimalSum() {
+        Assert.assertEquals(new BigDecimal("11"), PrecisionUtil.sum(null,
+                BigDecimal.ZERO, BigDecimal.ONE, BigDecimal.TEN));
+        Assert.assertEquals(BigDecimal.ONE,
+                PrecisionUtil.sum(BigDecimal.ONE, null));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Add test with no `null`s

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    @Test
+    public void testStringSum() {
+        Assert.assertEquals("1", PrecisionUtil.sum("1", null));
+        Assert.assertEquals("6", PrecisionUtil.sum(null, "1", "2", "3"));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:94]: Add test with no `null`s

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    @Test
+    public void testBigDecimalRound() {
+        Assert.assertEquals(new BigDecimal("3"),
+                PrecisionUtil.round(new BigDecimal("2.567"), 0));
+        Assert.assertEquals(new BigDecimal("2.6"),
+                PrecisionUtil.round(new BigDecimal("2.567"), 1));
+        Assert.assertEquals(new BigDecimal("2.57"),
+                PrecisionUtil.round(new BigDecimal("2.567"), 2));
+        Assert.assertEquals(new BigDecimal("2.567"),
+                PrecisionUtil.round(new BigDecimal("2.567"), 3));
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:156]: Add test with precision `4` (but still using `2.567`)

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 11 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:156]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 1 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:143]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 2 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:132]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 3 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:125]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 10 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:94]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 9 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 8 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java:77]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 6 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 7 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 5 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 4 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Done

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                PrecisionUtil.subtract(null, BigDecimal.TEN));
+        Assert.assertEquals(BigDecimal.ONE,
+                PrecisionUtil.subtract(BigDecimal.ONE, null));
+        Assert.assertEquals(new BigDecimal("3"),
+                PrecisionUtil.subtract(new BigDecimal("7"), new BigDecimal("4")));
+    }
+
+    @Test
+    public void testBigDecimalMultiply() {
+        Assert.assertEquals(new BigDecimal("10"),
```
Comment (freddytuxworth) [inline src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: Use `BigDecimal.TEN`

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

File: src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java (hunk 12 above)
Comment (Braavos6) [reply src/test/java/com/github/neatlife/jframework/util/PrecisionUtilTest.java]: oops.. missed that one :)

**Review by freddytuxworth — APPROVED** [review]
